synapse/handlers/federation_event.py

```diff
@@ -955,6 +960,20 @@ async def _compute_event_context_with_maybe_missing_prevs(
         seen = await self._store.have_events_in_timeline(prevs)
         missing_prevs = prevs - seen
 
+        # If we've already recently attempted to pull this missing event recently, don't
+        # try it again so soon. Since we have to fetch all of the prev_events, we can
+        # bail early here if we find any to ignore.
+        prevs_to_ignore = await self._store.filter_event_ids_with_pull_attempt_backoff(
+            room_id, missing_prevs
+        )
+        if len(prevs_to_ignore) > 0:
+            raise FederationError(
+                "ERROR",
+                429,
+                "Not attempting to pull event that we already tried to pull recently (backing off).",
+                affected=prevs_to_ignore[0],
+            )
```

Comment:
Sytest failure in CI:

https://github.com/matrix-org/synapse/actions/runs/3185784983/jobs/5198427610
```
FAILURE: #588: Can backfill purged history
```

But it's also so painful to look into it locally since it takes 4 minutes to run a single Sytest test. Created a separate issue to track the slowness there, https://github.com/matrix-org/sytest/issues/1299

Reply:
Right now the test is failing because no server has the invite event when `server-0` invited `server-1`.

 1. `server-0` invites user from `server-1`
 1. `server-1` receives invite and persists it as an outlier
 1. `server-1` joins the room
 1. `server-0` purges all of their room history
 1. Now when `server-0` tries to backfill from `server-1` and asks about the invite, `server-1` just throws up it's hands and says `State not known at event` because they only know it as an outlier
 
https://github.com/matrix-org/synapse/blob/dcced5a8d76b94e372aefa7d1f05ec0dbc22ea0d/synapse/handlers/federation.py#L1180-L1184

---

I would expect `server-1` to actually backfill the invite event at some point.

Reply:
Could be https://github.com/matrix-org/synapse/issues/9864 / https://github.com/matrix-org/sytest/issues/1033

Reply:
I was able to make things pass by forcing the remote server (`server-1`) to backfill all of the historical state via a `/messages` request. This way everything is de-outliered and ready for `server-0` to ask about again after the purge. Fixed in Sytest: https://github.com/matrix-org/sytest/pull/1300

Reply:
The better way forward seems to be to not to cause the cascade of failed pull attempts in the first place, see https://github.com/matrix-org/synapse/pull/13816#discussion_r987676306